### PR TITLE
CDK: Add schema inferrer class

### LINKED ISSUE #20819: [Low-Code CDK] Implement schema inferrer
As specified in the tech spec, there should be a helper class in the CDK which can automatically create schemas based on sample records.

## Acceptance criteria

* There is a util class exposed as part of the CDK which can be passed records for multiple streams and will return the derived schemas

### PR DESCRIPTION
**Pull request by flash1293**
Fixes #20819 

## What

This is the first part of adding automatic schema inferring to the connector builder UI. It adds a utility class `SchemaInferrer` to the CDK which can be used by the connector builder server and later on directly through the `read` command of a declarative source to infer the schema from some records of test data.

## How

Use the same lib powering https://github.com/airbytehq/airbyte/tree/master/tools/schema_generator to implement this generic class (do not use it anywhere yet). In a subsequent PR this class is used in the connector builder server to pipe all encountered records through and send the derived schema along with the response to the connector builder UI.

### CODE REVIEW

File: airbyte-cdk/python/CHANGELOG.md
Comment (brianjlai) [inline airbyte-cdk/python/CHANGELOG.md:3]: nit: but given this is additive non-breaking functionality (and not a bug fix), we should bump the minor version so `0.17.0`

File: airbyte-cdk/python/airbyte_cdk/utils/__init__.py
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/utils/__init__.py:4]: need an extra space between the licence and import or our CI formatting checks will fail

File: airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py:71]: The original tech spec outlined the `accumulate` and `get_inferred_schemas` functions, but for context what's the use case for a single stream or how are we using this `get_stream_schema()` function
Also why do we no longer need the `reset()`?

**Review by brianjlai — APPROVED** [review]
no blocking issues but just a few nits about formatting and version. Also wanted to get some context on interface changes from the original tech spec [here](https://docs.google.com/document/d/1fhAIHYBoqUk5vtTqEqcNrQMoXrUwrZ9HtAB5asgy74k/edit#).

I think the CI checks should pass once you fix formatting, but double check there aren't any other issues before the publish + merge.

File: airbyte-cdk/python/unit_tests/utils/test_schema_inferrer.py
Comment (sherifnada) [inline airbyte-cdk/python/unit_tests/utils/test_schema_inferrer.py:13]: could we also test for arrays and objects?

File: airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py
Comment (sherifnada) [reply airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py:71]: fwiw when writing the spec i didn't account for this object needing to infer schema on multiple streams (that iface assumed a single stream). I think this iface makes more sense.

File: airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py:48]: you can simplify the `if/else` block of `accumulate` by using `defaultdict` e.g: 

```
def __init__():
  self.builders = defaultdict(NoRequiredSchemaBuilder)

...

def accumulate(self, record): 
  self.builders[record.stream].add_object(record.data)

```

File: airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py:21]: are these two if blocks what prevent generating `required` fields?

**Review by sherifnada — CHANGES_REQUESTED** [review]

File: airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py
Comment (flash1293) [reply airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py:71]: Yeah, when integrating with the API I noticed that for that use case we only need a specific one and this API seemed nicer. The `get_inferred_schemas` will still be nice for the planned integration with the `read` command / implementing a separate command for schema inferrence

File: airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py
Comment (flash1293) [reply airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py:48]: Neat, didn't know about defaultdict

File: airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py
Comment (flash1293) [reply airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py:21]: In a sense, I copied this over from https://github.com/airbytehq/airbyte/blob/master/tools/schema_generator/schema_generator/infer_schemas.py#L41

However on checking the genson code I changed the logic to just call the existing `to_schema` logic of the object strategy and then dropping the `required` prop. It's doing a very small amount of additional work but it seems much cleaner, LMK what you think.

File: airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py
Comment (flash1293) [reply airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py:71]: I dropped the `reset` because in the places where we plan to use this class it's much easier to just lose the reference to the inferrer instance and have the runtime clean it up, then create a new instance for the next usage. Happy to change that, but it doesn't seem heavy enough to explicitly keep an instance or a pool of them around.

File: airbyte-cdk/python/unit_tests/utils/test_schema_inferrer.py
Comment (sherifnada) [inline airbyte-cdk/python/unit_tests/utils/test_schema_inferrer.py:47]: wouldn't it be better to compare without casting to string to avoid mismatches around things like ordering of properties, which doesn't matter in objects? e.g would it be better to do 
```
assert (
inferrer.get_inferred_schemas()["my_stream"] 
== {"$schema": "http://json-schema.org/schema#", "type": "object", "properties": {"field_A": {"type": ["number", "string"]}}}
```

File: airbyte-cdk/python/unit_tests/utils/test_schema_inferrer.py
Comment (sherifnada) [inline airbyte-cdk/python/unit_tests/utils/test_schema_inferrer.py:51]: would recommend considering the use of `pytest.mark.fixture` here to parametrize tests

**Review by sherifnada — APPROVED** [review]
LGTM once the comment is addressed, no need for another review

**Pull request merged**
